sound/soc/sof/intel/hda-dsp.c

```diff
@@ -825,15 +825,31 @@ void hda_dsp_d0i3_work(struct work_struct *work)
 	struct hdac_bus *bus = &hdev->hbus.core;
 	struct snd_sof_dev *sdev = dev_get_drvdata(bus->dev);
 	struct sof_dsp_power_state target_state;
+	struct snd_sof_client *client;
+	struct device *dev;
+	bool remain = false;
 	int ret;
 
 	target_state.state = SOF_DSP_PM_D0;
 
-	/* DSP can enter D0I3 iff only D0I3-compatible streams are active */
-	if (snd_sof_dsp_only_d0i3_compatible_stream_active(sdev))
-		target_state.substate = SOF_HDA_DSP_PM_D0I3;
-	else
+	/*
+	 * DSP can enter D0I3 iff only D0I3-compatible streams are active.
+	 * Check all clients for this condition.
+	 */
+	list_for_each_entry(client, &sdev->client_list, list) {
+		dev = &client->pdev->dev;
+
+		if (client->type != SOF_CLIENT_AUDIO)
+			continue;
+
+		if (!snd_sof_dsp_only_d0i3_compatible_stream_active(dev))
+			remain = true;
```

Comment:
I am not a big fan of this, this looks super-racy.
We'd probably need some sort of locking, but it'd be even better to have notifications with a mutex that changes the internal state.